### ISSUE TITLE
Clean up Composer, be mindful of performance and maintaining existing behavior

### ISSUE THREAD

**Issue by mchandschuh** [QuantConnect/Lean#3836]
Another thing I'm wondering is  that `Composer.Reset()` was changed to **not** perform a reset. This seems misleading. I haven't changed that as my goal was to not change any behavior, but I'm thinking we should minimally make `Reset()` actually perform a reset again and then (if required) add a separate method that simply clears the `_exportedValues` cache -- but I suspect the reason `Reset()` was changed was to improve test speeds, in which case, perhaps we just don't call `Reset()` -- or again, if required, we could add another method that does what we it it in an appropriately named method.

Relevant commits:
> `e4fc00efbf876e39280344187017d707c325ca56`
> `9865fe63e0da90c85ba26ecd4fb0067b7bbab825`

#### Expected Behavior

Code should generally be kept clean and readable.

#### Actual Behavior

Over time, as is with all things, the code in `Composer.cs` has gotten a little unwieldy, making it hard to understand exactly what it's doing without really digging into it.

#### Potential Solution

Perform a refactor but be extremely mindful of performance as most of the recent changes in `Composer.cs` are aimed at improving LEAN startup times and unit test times.

#### Reproducing the Problem

Run LEAN. For timing, I used the following code in `Program.cs`'s `Main()` function:
<details><summary>Code snippet to reproduce timings used for this PR</summary>
``` csharp
static void Main(string[] args)
{
    var start = DateTime.UtcNow;

    //Initialize:
    var mode = "RELEASE";
    #if DEBUG
        mode = "DEBUG";
    #endif

    if (OS.IsWindows)
    {
        Console.OutputEncoding = System.Text.Encoding.UTF8;
    }

    // expect first argument to be config file name
    if (args.Length > 0)
    {
        Config.MergeCommandLineArgumentsWithConfiguration(LeanArgumentParser.ParseArguments(args));
    }

    var environment = Config.Get("environment");
    var liveMode = Config.GetBool("live-mode");
    Log.DebuggingEnabled = Config.GetBool("debug-mode");
    Log.LogHandler = Composer.Instance.GetExportedValueByTypeName<ILogHandler>(Config.Get("log-handler", "CompositeLogHandler"));

    //Name thread for the profiler:
    Thread.CurrentThread.Name = "Algorithm Analysis Thread";
    Log.Trace("Engine.Main(): LEAN ALGORITHMIC TRADING ENGINE v" + Globals.Version + " Mode: " + mode + " (" + (Environment.Is64BitProcess ? "64" : "32") + "bit)");
    Log.Trace("Engine.Main(): Started " + DateTime.Now.ToShortTimeString());

    //Import external libraries specific to physical server location (cloud/local)
    LeanEngineSystemHandlers leanEngineSystemHandlers;
    try
    {
        leanEngineSystemHandlers = LeanEngineSystemHandlers.FromConfiguration(Composer.Instance);
    }
    catch (CompositionException compositionException)
    {
        Log.Error("Engine.Main(): Failed to load library: " + compositionException);
        throw;
    }

    //Setup packeting, queue and controls system: These don't do much locally.
    leanEngineSystemHandlers.Initialize();

    //-> Pull job from QuantConnect job queue, or, pull local build:
    string assemblyPath;
    var job = leanEngineSystemHandlers.JobQueue.NextJob(out assemblyPath);

    if (job == null)
    {
        throw new Exception("Engine.Main(): Job was null.");
    }

    LeanEngineAlgorithmHandlers leanEngineAlgorithmHandlers;
    try
    {
        leanEngineAlgorithmHandlers = LeanEngineAlgorithmHandlers.FromConfiguration(Composer.Instance);
    }
    catch (CompositionException compositionException)
    {
        Log.Error("Engine.Main(): Failed to load library: " + compositionException);
        throw;
    }

    var end = DateTime.UtcNow;
    var elapsed = end - start;
    Log.Error($"COMPOSER INITIALIZATION/RESOLUTION TIME: {elapsed.TotalMilliseconds:0.00}ms");
```
</details>

#### Checklist

- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [x] I have provided detailed steps to reproduce the issue

**Comment by mchandschuh** [comment]
I've pushed the `Program.cs` file I used for timing to a [branch](https://github.com/mchandschuh/Lean/tree/refactor-3836-clean-up-composer-and-improve-lean-startup-performance.timing-code). It would be great if someone else could benchmark to confirm the improvement. Here's the timings I got:
```
PR: 345.00ms (cold start)
    231.99ms
    201.04ms
    191.00ms
    
MASTER: 1603.00ms (cold start)
    346.00ms
    355.98ms
    349.04ms
```

**Comment by jaredbroad** [comment]
Merged in #3846
